File: go/day3/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

func main() {
	input := read("input/day3/sample.txt")
	fmt.Println(input)
	fmt.Println(puzzle(input, findLargestJoltage1))
	fmt.Println(puzzle(input, findLargestJoltage2))
}

func puzzle(banks [][]int, fn func([]int) int) int {
	sum := 0
	for _, bank := range banks {
		sum += fn(bank)
	}
	return sum
}

func findLargestJoltage1(bank []int) int {
	high, low := bank[0], bank[1]
	for i := 1; i < len(bank); i++ {
		if i < len(bank)-1 && bank[i] > high {
			high, low = bank[i], bank[i+1]
		} else if bank[i] > low {
			low = bank[i]
		}
	}
	return high*10 + low
}

func findLargestJoltage2(bank []int) int {
	digits := _findLargestJoltage2(bank, 12, make([]int, 0, 12))
	sum := 0
	for pos, n := range digits {
		sum += n * int(math.Pow10(11-pos))
	}
	return sum
}

func _findLargestJoltage2(bank []int, size int, current []int) []int {
	if size == 0 {
		return current
	}
	high, index := bank[0], 0
	for i := 1; i <= len(bank)-size; i++ {
		if bank[i] > high {
			high, index = bank[i], i
		}
	}
	return _findLargestJoltage2(bank[index+1:], size-1, append(current, high))
}

func _max(digits []int) int {
	high := digits[0]
	for _, n := range digits[1:] {
		if n > high {
			high = n
		}
	}
	return high
}

func read(file string) [][]int {
	f, err := os.Open(file)
	if err != nil {
		panic(err)
	}

	s := bufio.NewScanner(f)
	result := [][]int{}
	for s.Scan() {
		row := []int{}
		for _, char := range s.Text() {
			row = append(row, (int(char - '0')))
		}
		result = append(result, row)
	}
	return result
}

File: python/day3/__main__.py
def read(filename):
    with open(filename) as f:
        return [ [ int(char) for char in row[:-1] ] for row in f ]

def find_largest1(bank):
    high, low = bank[0], bank[1]
    for i in range(1, len(bank)):
        if i < len(bank) - 1 and bank[i] > high:
            high, low = bank[i], bank[i+1]
        elif bank[i] > low:
            low = bank[i]
    return high*10 + low

def find_largest2(bank):
    digits = _find_largest2(bank, size=12, digits=[])
    sum = 0
    for pos in range(len(digits)):
        n = digits[pos]
        sum += n*(10**(11-pos))
    return sum

def _find_largest2(bank, size=12, digits=[]):
    if size == 0:
        return digits
    if size == 1:
        digits.append(max(bank))
        return digits
    high, index = bank[0], 0
    for i in range(len(bank[1:-size+1])):
        if bank[i+1] > high:
            high, index = bank[i+1], i+1
    digits.append(high)
    # print(bank, bank[:-size+1], high, index, size)
    return _find_largest2(bank[index+1:], size-1, digits)

def puzzle(banks, fn):
    sum = 0
    for bank in banks:
        sum += fn(bank)
    return sum

data = read("input/day3/sample.txt")
# print(data)
print(puzzle(data, find_largest1))
print(puzzle(data, find_largest2))
# print(find_largest2(data[1]))


# 3121910778619
# 3121190777711


# 987654321111
# 811111111119
# 434234234278
# 888911112111

# 987654321111
# 811111111119
# 434234234278
# 888911112111